highly_variable_gene batch_key variant fix

LuckyMD opened the pull request:
This PR addresses part of #758. I just added a gene filtering addition to the `batch_key` variant of `sc.pp.highly_variable_genes()`. This ensures that the function does not fail because a gene is not expressed in a batch.

I would welcome some feedback on setting dispersions to 0 for filtered out genes. I think this is the standard set by Seurat, and also what is implemented by @gokceneraslan in the original function. 

I have tested on my own data for correct implementation. Testing for result is more difficult in this I guess...

LuckyMD committed:
added gene filtering to highly_variable_gene batch_key variant

LuckyMD committed:
fixed whitespace issue

flying-sheep committed:
Style

flying-sheep reviewed (changes requested):
I think this could be done more efficiently by using the index returned from `filter_genes(..., inplace=False)` in `_highly_variable_genes_single_batch` and instead of the whole data frame merging you add in the current version of your changes.

LuckyMD commented:
> I think this could be done more efficiently by using the index returned from `filter_genes(..., inplace=False)` in `_highly_variable_genes_single_batch` and instead of the whole data frame merging you add in the current version of your changes.

I guess that would depend if you want to have a `filter_genes` call in the HVG selection function every time, or whether you only want it in there in a case where `filter_genes` normally doesn't work. You typically use it on the whole dataset, but not per batch.

Another issue atm is that if you set the verbosity high, then the `filter_genes()` call gives you an output, which is not really intended as the user can't see the function call.

ivirshup commented on scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -250,25 +250,52 @@ def highly_variable_genes(
+
+            # Filter to genes that are in the dataset
+            genes_before = adata_subset.var_names
+            filter_genes(adata_subset, min_cells=1)
```
This makes a copy of the whole `AnnData`, which isn't needed here. I'd follow @flying-sheep's recommendation to use `inplace=False` or calculate it directly. You can probably hide the logging output with a context manager if you want.

flying-sheep replied:
There’s only one for filtering warnings, so I quickly built one:

https://github.com/theislab/scanpy/blob/e63c2df73fc324675b5cff52b444101852efe055/scanpy/_settings.py#L39-L46

Usage:

```py
with sc.settings.verbosity.override(sc.Verbosity.error):
    sc.noisy_function(adata)
```

LuckyMD replied:
Thanks! Used that now.

ivirshup commented on scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -250,25 +250,52 @@ def highly_variable_genes(
+            tmp = hvg.set_index('gene')
+            tmp = tmp.loc[genes_before]
+            tmp['gene'] = tmp.index
+            tmp.index = hvg.index
```
Do these lines end up doing anything?

LuckyMD replied:
They didn't before... I had forgotten to change the line after (`df.append(hvg)` is now `df.append(tmp)`). Thanks for the heads up. As I mentioned, I have only tested that it runs so far.

ivirshup commented on scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -250,25 +250,52 @@ def highly_variable_genes(
+            genes_missing = genes_before.difference(genes_after)
+
+            hvg = _highly_variable_genes_single_batch(
+                adata_subset,
```
This function should probably take an array/ sparse matrix directly, since it doesn't actually use anything else from the `AnnData`.

LuckyMD replied:
Hmm... I didn't make that function. That was done by @gokceneraslan when he added `batch_key`. I would therefore be inclined to leave it like that in this PR.

ivirshup commented:
Am I correct in thinking this is a fix for a batch having zero variance for a gene? What was the output in that case?

LuckyMD commented:
This is mainly a fix for cases when multiple genes have zero variance. The output in that case was that `sc.pp.highly_variable_genes()` failed as the bin boundaries were too close to one another. The error I got is:
```
ValueError: Bin edges must be unique: array([          -inf, 9.99999996e-13, 9.99999996e-13, 3.71624832e-03,
       4.50723944e-03, 5.04237041e-03, 7.96065722e-03, 9.17631686e-03,
       1.15813100e-02, 1.34968273e-02, 1.62843971e-02, 1.89858746e-02,
       2.27864407e-02, 2.76163086e-02, 3.43018658e-02, 4.27573830e-02,
       5.52219763e-02, 7.87758350e-02, 1.42211060e-01, 3.10728383e+00,
                  inf]).
You can drop duplicate edges by setting the 'duplicates' kwarg
```

I figured the best way forward would be to exclude those genes from the function, rather than changing the bins in the `_highly_variable_genes_single_batch()` function with the `duplicates` argument as suggested in the error.

LuckyMD committed:
fixed bug

ivirshup commented:
> This is mainly a fix for cases when multiple genes have zero variance.

Could you add that as the test case? When some genes aren't expressed in a batch you won't get an error.

> the best way forward would be to exclude those genes from the function

I think the approach of masking out the non-expressed genes sounds reasonable, since that's what you'd probably do if it were just one dataset. I'd definitely defer to @gokceneraslan on any more about the appropriateness of the approach.

LuckyMD committed:
Merge branch 'master' of https://github.com/theislab/scanpy into hvg_batch_fix

LuckyMD committed:
added test, improved code efficiency, used context manager

LuckyMD committed:
bug fix testing hvg function

LuckyMD committed:
another bug fix testing hvg function

LuckyMD committed:
final bug fix testing hvg function

LuckyMD commented:
@ivirshup Is this fine for you as is? I think I've addressed all the points now. I'd be keen for this to be merged asap as we need this for another project where this code is failing.

Note that I have not addressed the issue with `inplace=False` when using the `batch_key` parameter.

LuckyMD commented:
I had to use the pbmc3k dataset for testing, as the error doesn't occur on blobs or pbmc68k_reduced. To test I need sufficient genes that have 0 variance in a subset of the cells.

LuckyMD committed:
test with pbmc68k

LuckyMD commented:
Cancel that @flying-sheep sheep helped me find a way around to test with `pbmc68k_reduced`. This should speed up Travis again.

flying-sheep commented:
Works for me! I’d say we refactor the helper function in a separate PR
